### DISCUSSION TITLE
逐个获取page.chars方法无法得到有些table里的chars,格式一模一样的table有的就不能获取

### DISCUSSION THREAD

**Discussion by zyc1128** [jsvine/pdfplumber#1150] [Ask for help with specific PDFs]
## Describe the bug

*A clear and concise description of what the bug is.*

## Have you tried [repairing](https://github.com/jsvine/pdfplumber/blob/stable/docs/repairing.md) the PDF?

*Please try running your code with `pdfplumber.open(..., repair=True)` before submitting a bug report.*

## Code to reproduce the problem

*Paste it here, or attach a Python file.*

## PDF file

*Please attach any PDFs necessary to reproduce the problem.*

*If you need to redact text in a sensitive PDF, you can run it through [JoshData/pdf-redactor](https://github.com/JoshData/pdf-redactor).*

## Expected behavior

*What did you expect the result __should__ have been?*

## Actual behavior

*What actually happened, instead?*

## Screenshots

*If applicable, add screenshots to help explain your problem.*

## Environment

- pdfplumber version: [e.g., 0.5.22]
- Python version: [e.g., 3.8.1]
- OS: [e.g., Mac, Linux, etc.]

## Additional context

*Add any other context/notes about the problem here.*

**Comment by jsvine** [comment]
Hi @zyc1128, it sounds like you need help with a specific PDF. Can you share the PDF here, along with the code you're using, and a longer explanation of what difficulty you're having?

**Reply by zyc1128** [reply]
with pdfplumber.open("d:/产品工艺文件/芬兰橱柜.pdf") as doc:

    texts=""
    
    for page in doc.pages:

        obj = page.chars
        img = page.images
        for e in range(0, len(obj), 1):
            my_char = obj[e]
            my_char_ctm = CTM(*my_char["matrix"])

            if my_char_ctm.skew_x == 0.0:
                texts = texts + obj[e]["text"]
            else:


              with open('d:/测试.txt', 'a') as f:

                q=str(my_char_ctm.skew_x)
                f.write("data: " + obj[e]["text"] + "参数: " + q + "页码："+str(page.page_number)+"\n")
    print(texts)
我逐字获取内容，正常字体合并成texts并打印，不正常的写入文本，最后发现第三页的产品描述的那张表texts里面没有，txt里也没有

**Reply by zyc1128** [reply]
[Uploading 芬兰橱柜.pdf…]()

**Reply by jsvine** [reply]
Unfortunately, this upload does not seem to have worked; it's just a link to this page.

**Reply by zyc1128** [reply]
> Unfortunately, this upload does not seem to have worked; it's just a link to this page.

upload below again,dose this work?

**Reply by jsvine** [reply]
Yes, that upload worked, thanks. Re. your question about tables, when I run this code:

```python
# To remove the watermarking text
filtered = pdf.pages[2].filter(lambda obj: not (obj.get("fontname") == "MWRYEA+NSimSun"))

for i, table in enumerate(filtered.extract_tables()):
    print(f"=== Table {i} ===")
    print(table)
    print("\n---\n")
```

... I get this:

```
=== Table 0 ===
[['款式', '场景图', '拉手示意图', '备注'], ['芬兰', '', '', '1.门板、装饰板：18mm厚，平板无造型，四周1.3mm同色PVC封边；\n2.地柜可配合潘多拉铝线箱体使用；\n（CB365/366颜色可选氧化色、巧克力或铜拉丝色）；'], ['芬兰II', '', '', '1.门板：18mm厚，平板无造型，三边1.3mm同色PVC封边，拉手边\n0.45mm巧克力铝色（FL120系列拉手）/氧化铝色PVC封边（FL033\n系列拉手）；\n2.装饰板：18mm厚，平板无造型，四周1.3mm同色PVC封边；']]

---

=== Table 1 ===
[['门板纹理示意', '板材最大规格', '门板纹理', '装饰板纹理'], ['', '1200*2400*18mm\n纹理方向为2400mm\n长度方向；', '门板纹理：横向/竖向纹理；\n当门板高度＞1200mm时，做横向纹\n理门板时必须为复合门板；\n当门板高度≤1200mm时，为横向/竖\n向纹理；', '装饰板纹理：横向/竖向纹理；\n当装饰板高度＞1200mm时，做横向纹理时需做多\n块装饰板拼接使用；\n当装饰板高度≤1200mm时，为横向/竖向纹理；'], ['备注：\n1. 部分板材装饰板最大规格：1200*2700*18mm,2700长度方向，可做整板装饰板，材质如下（共11种）：\n特色装饰板A类：浮雕暖白（SD042）、浮雕胡桃（SD043）、浮雕新深灰橡（SD044）、浮雕新黑橡（SD045）、浮雕水洗橡（SD047）\n浮雕沙贝文（SD048）、浮雕灰褐橡（SD050）；\n特色装饰板B类：同步木纹白橡（SD033）、同步木纹本色橡（SD034）、苍榆木（SD040）、浮雕烟熏橡（SD046）；\n2. 特色装饰板B类-白木纹门板表面有两种方向纹理，一是印刷木纹纹理2400长度方向（深浅木纹，且有木纹肌理），二是不规则\n锯齿纹纹理1200mm长度方向；两种纹理互相垂直,以印刷木纹纹理为准）；\n3. 特色装饰板B类-深咖啡石纹双面均匀凹凸石纹纹理，门板与装饰板无纹理要求；', None, None, None]]

---

=== Table 2 ===
[['款式', '箱体花色', '层板铝包边', '铰链', '拉手', '踢脚板'], ['芬兰', '博洛尼\n正常箱体花色', '巧克力色', 'Blum一体化环保\n型内置阻尼铰链', 'C级拉手', '铝踢脚板\n氧化色/巧克力色'], ['芬兰II', None, None, None, 'FL120/FL033系列拉手', None]]

---
```

Does that work for you?

**Reply by zyc1128** [reply]
谢谢，很有帮助！我用了你这个示例的filtered.extract_text()：
 texts = ""
 for text in enumerate(filtered.extract_text()):
     texts=texts+text[1]
 print(texts)
提取了文字和table,这个texts是一个类似于string的结构，如何让这里面table内容以类似于dataframe的表现形式存在于这个texts中？就像原PDF一样，看上去就是去掉了table里的横线和竖线？

**Reply by zyc1128** [reply]
我用一样的代码提取另一个PDF的文字，它又把一些
[Uploading JS-B5-04-026D-博洛尼家具-新伯尔尼亚系列型录-V1.2.pdf…]()
没用的斜的字提取出来了

**Reply by zyc1128** [reply]
[JS-B5-04-026D-博洛尼家具-新伯尔尼亚系列型录-V1.2.pdf](https://github.com/user-attachments/files/15861078/JS-B5-04-026D-.-.-V1.2.pdf)

**Reply by jsvine** [reply]
To extract tables, you'll want to use the `.extract_table(...)` or `.extract_tables(...)` methods. To convert those into DataFrames, you can pass the nested-list representation of the table to `pandas.DataFrame(...)`. E.g.:

```python
import pandas as pd
import pdfplumber

pdf = pdfplumber.open("path.pdf")
page = pdf.pages[0]
table = page.extract_table()
df = pd.DataFrame(table)
```

**Comment by zyc1128** [comment]
[芬兰橱柜.pdf](https://github.com/user-attachments/files/15796509/default.pdf)

**Reply by jsvine** [reply]
When I click that link, I get this:

<img width="93" alt="Screenshot 2024-06-12 at 4 28 22 PM" src="https://github.com/jsvine/pdfplumber/assets/534702/b3948e8d-a2ca-43a2-b678-028ec253c392">

**Reply by zyc1128** [reply]
[芬兰橱柜.pdf](https://github.com/user-attachments/files/15812426/default.pdf)
